docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+The object descriptor layout is still likely to change, the current layout is:
+
+64-bit integer size.
+Pointer to a trace function, used in traced regions to walk reachable objects.
+Pointer to a finalise function, called when an object is dead but before other objects in the same region are deleted (though they may have already been finalised.
+Pointer to a notify function, invoked when an object is notified of a noticeboard event.
+Pointer to a destructor, which is called after some objects in a region have been deallocated and may not follow any `mut` pointers.
+The remainder of the descriptor is the vtable.
+
+Methods in the vtable use *selector colouring*.
+Each selector (method name and types) is assigned a separate vtable index.
+Pairs of selectors that do not appear in the same interface are then combined to use the same index.
+Selector colouring is not essential but can dramatically reduce the size of vtables.
+
+*Note:* If we can distinguish between sentinel objects and others at object-creation time, the region metadata field for non-sentinel objects could be a 32-bit value that is either an integer or an index into a table, rather than a pointer.
+This would be a noticeable saving on CHERI systems and a small saving on 64-bit systems, at the cost of an extra indirection for some memory-management operations.
+We could also use a 32-bit index into a descriptor table, as the Fuchsia C++ ABI does: Programs that have more than 2^32 distinct types arise only as a result of poor life choices.
+
+*Note:* The above choices are visible to the runtime but are part of the internal ABI and so it would be possible to have size and performance optimised variants available on a per-program basis.
+
+*Note:* On AArch64 systems where the top-byte-ignore feature is available, using the top bits in the descriptor pointer instead of the bottom ones would improve efficiency of descriptor accesses.
+
+### Field layout
+
+In C, `struct` layout is a trade between storage density and accidental aliasing.
+If two fields are accessed concurrently, false sharing in the cache can cause significant performance bottlenecks.
+In Verona; however, no mutable objects are ever shared and so, aside from reference count manipulations on immutable objects and interactions between adjacently allocated objects in different regions, there are no negative interactions with modern caches.
+The Verona field layout algorithm is therefore free to optimise for space.
+
+All fields whose size is a power-of-two should be sorted by size and laid out largest to smallest.
+The end of each section is more strongly aligned than the next requires and so this does not require any padding.
+
+Next, any union types that have a power-of-two sized payload but a smaller discriminator should be split into payload and discriminator.
+The payloads should be inserted into the sections for the same sized power-of-two objects.
+The discriminators should be packed into a single larger value.
```

Comment:
Taking those in opposite order:

For (2), anything with descriptors embedded in the payload, the result should be a power-of-two type and so the first rules apply.

In think (1) is a much more interesting question and one that I hadn't fully considered.  It probably does make sense to float them backwards because then we can always do an aligned load + ubfx to get any field's descriptor.